util/topgen/resets.py

```diff
@@ -0,0 +1,188 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Dict, Optional
+from .clocks import Clocks
+
+
+class ResetItem:
+    '''Individual resets'''
+    def __init__(self, hier: Dict[str, str], raw: Dict[str, object], clocks: Clocks):
+        if not raw['name']:
+            raise ValueError('Reset has no name')
+
+        self.name = raw['name']
+        self.gen = raw.get('gen', True)
+        self.rst_type = raw.get('type', 'top')
+
+        self.path = ""
+        if self.rst_type == 'top':
+            self.path = f"{hier['top']}rst_{self.name}_n"
+        elif self.rst_type == 'ext':
+            self.path = f"{hier['ext']}{self.name}"
+
+        # to be constructed later
+        self.domains = []
+        self.shadowed = False
+
+        self.parent = raw.get('parent', "")
+
+        # This can be a source clock or a derived source
+        if self.rst_type != 'ext':
+            self.clock = clocks.get_clock_by_name(raw['clk'])
+        else:
+            self.clock = None
+
+        self.sw = bool(raw.get('sw', 0))
+
+    def _asdict(self) -> Dict[str, object]:
+        ret = {
+            'name': self.name,
+            'gen': self.gen,
+            'type': self.rst_type,
+            'domains': self.domains,
+            'shadowed': self.shadowed,
+            'parent': self.parent,
+            'clock': "" if self.clock else "",
```

Comment:
I'm not sure that's what you meant: you're adding an empty string unconditionally :-)

Unfortunately, the only way that I know to do this is to drop the field here and then add something like this below:
```python3
      if self.clock:
           ret['clock'] = self.clock
```
:-(

Also `self.sw` needs converting back from `bool` to `int` on the next line. For example, `spi_device` has gone from `sw: 1 to `sw: true` in the generated hjson (or we change the hjson schema and the parsing code to use bools: either works for me!)

Reply:
oops that one was a typo.  I meant to do `self.clock` instead of an empty string.
yeah i do think i prefer booleans everywhere.. let me do that.